File: 2019/day6/day6.py
from collections import deque
f = open('input', 'r')

lines = f.read().splitlines()

class node:
    def __init__(self, n):
        self.name = n
        self.children = []
        self.parent = None

    def addchild(self, child):
        self.children += [child]
        child.parent = self

    def setlevel(self, i):
        self.level = i

def dfs(n, level):
    n.setlevel(level)
    for child in n.children:
        dfs(child, level+1)

def bfs(src, dest):
    cur = (src, 0)
    d = deque()
    v = set()
    v.add(cur[0].name)
    while not cur[0] == dest:
        if (not cur[0].parent == None) and (cur[0].parent.name not in v):
            d.append((cur[0].parent, cur[1]+1))
            v.add(cur[0].parent.name)
        for child in cur[0].children:
            if child.name not in v:
                d.append((child, cur[1]+1))
                v.add(child.name)
        cur = d.popleft()
    return cur[1]

nodes = {}

for line in lines:
    orbit = line.split(')')
    if not orbit[0] in nodes:
        nodes[orbit[0]] = node(orbit[0])
    if not orbit[1] in nodes:
        nodes[orbit[1]] = node(orbit[1])
    nodes[orbit[0]].addchild(nodes[orbit[1]])

dfs(nodes['COM'], 0)
res = 0
for name in nodes:
    res += nodes[name].level

print(res)
print(bfs(nodes['YOU'], nodes['SAN'])-2)

File: 2019/day6/day6.cpp
#include <iostream>
#include <queue>
#include <stdio.h>
#include <string>
#include <unordered_map>
#include <vector>

struct node {
    node() {
        node("");
    }

    node(const std::string& n) : name(n) {}

    void addchild(node& child) {
        children.push_back(&child);
        child.parent = this;
    }

    void setlevel(int lev) {
        level = lev;
    }

    std::string name;
    std::vector<node*> children;
    node* parent;
    int level = 0;
};

void dfs(node* n, int level) {
    n->setlevel(level);
    for (auto& child : n->children) {
        dfs(child, level + 1);
    }
}

int bfs(node* src, node* dest) {
    std::queue<node*> q;
    std::unordered_map<std::string, int> dists;
    dists[src->name] = 0;
    node* cur = src;
    while (cur->name.compare(dest->name) != 0) {
        if (cur->parent != nullptr && dists.find(cur->parent->name) == dists.end()) {
            q.push(cur->parent);
            dists[cur->parent->name] = dists[cur->name] + 1;
        }

        for (const auto& child : cur->children) {
            if (dists.find(child->name) == dists.end()) {
                q.push(child);
                dists[child->name] = dists[cur->name] + 1;
            }
        }

        cur = q.front();
        q.pop();
    }

    return dists[dest->name];
}

int main() {
    std::unordered_map<std::string, node> nodes;
    for (int i = 0; i < 1868; i++) {
        std::string x, y, line;
        std::cin >> line;
        x = line.substr(0, 3);
        y = line.substr(4);


        if (nodes.find(x) == nodes.end())
            nodes[x] = node(x);
        if (nodes.find(y) == nodes.end())
            nodes[y] = node(y);

        nodes[x].addchild(nodes[y]);
    }

    dfs(&(nodes["COM"]), 0);
    unsigned int res = 0;
    for (auto& n : nodes) {
        res += nodes[n.first].level;
    }

    printf("%u\n", res);

    int distance = bfs(&(nodes["YOU"]), &(nodes["SAN"]));
    printf("%d\n", distance - 2);
}